### PR TITLE
Update dependencies and docs for 4.1 release

### PR DESCRIPTION
**Pull request by MarkEWaite**
## Update docs and dependencies

Update the dependencies and the documentation page as preparation for eventual release of git plugin 4.1.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] Dependency or infrastructure update

### CODE REVIEW

**Comment by MarkEWaite** [comment]
@slonopotamus I would also like to update to parent pom 3.54, but when I attempt that, it fails to compile.  There seems to be a signature change in the jenkins test harness that I don't understand how to resolve.

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /home/mwaite/git/jenkins/git-plugin/src/test/java/org/jenkinsci/plugins/gittagmessage/AbstractGitTagMessageExtensionTest.java:[161,26] no suitable method found for buildAndAssertSuccess(J)
    method org.jvnet.hudson.test.JenkinsRule.<J,R>buildAndAssertSuccess(J) is not applicable
      (inference variable R has incompatible bounds
        equality constraints: R
        upper bounds: hudson.model.Run<J,R>,hudson.model.Queue.Executable)
    method org.jvnet.hudson.test.JenkinsRule.buildAndAssertSuccess(hudson.model.FreeStyleProject) is not applicable
      (argument mismatch; J cannot be converted to hudson.model.FreeStyleProject)
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  17.207 s
[INFO] Finished at: 2019-12-16T06:12:40-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project git: Compilation failure
[ERROR] /home/mwaite/git/jenkins/git-plugin/src/test/java/org/jenkinsci/plugins/gittagmessage/AbstractGitTagMessageExtensionTest.java:[161,26] no suitable method found for buildAndAssertSuccess(J)
[ERROR]     method org.jvnet.hudson.test.JenkinsRule.<J,R>buildAndAssertSuccess(J) is not applicable
[ERROR]       (inference variable R has incompatible bounds
[ERROR]         equality constraints: R
[ERROR]         upper bounds: hudson.model.Run<J,R>,hudson.model.Queue.Executable)
[ERROR]     method org.jvnet.hudson.test.JenkinsRule.buildAndAssertSuccess(hudson.model.FreeStyleProject) is not applicable
[ERROR]       (argument mismatch; J cannot be converted to hudson.model.FreeStyleProject)
```

**Comment by slonopotamus** [comment]
@MarkEWaite Whoops. Didn't think that that method could be proxied through other generics :(

[Here's a patch that fixes AbstractGitTagMessageExtensionTest](https://github.com/slonopotamus/git-plugin/commit/d343af62902356e2fa4a67e1eef03583a67f1aee). If you want a proper PR from it, it'll take some time.

**Comment by slonopotamus** [comment]
See #798, it fixes AbstractGitSCMSourceTest.when_commits_added_during_discovery_we_do_not_crash failure when run against newer JTH.

**Comment by slonopotamus** [comment]
I suspect that GitTagActionTest was broken even before this PR. But because assertion fails in the different thread, older hamcrest didn't properly mark test as failed.

**Comment by MarkEWaite** [comment]
The GitTagActionTest has a race condition that is sometimes visible.  I created the test while investigating [SECURITY-1095](https://jenkins.io/security/advisory/2019-01-28/#SECURITY-1095).  If it continues to be unreliable, I am very willing to delete the test.

The GitTagAction is a legacy operation in the plugin.  It creates a tag in a workspace after the build has completed.  The assumption is that the workspace exists after the build completes.  That assumption is not valid with ephemeral agents.  It does not push the tag anywhere, so the only repository which contains the tag is the repository inside the workspace.  Without pushing the tag anywhere, no other users will see the tag.

GitTagAction has no viable use case as far as I can tell.  It remains in the plugin because I don't want to spend the time or energy to delete the capability, document its deletion, and explain why the use case is not relevant.  It is easier to leave it as idle code than to remove it.

You can read the [diffs](https://github.com/jenkinsci/git-plugin/commit/f9152d943936b1c6b493dfe750d27f0caa7c0767) to see the types of horrors I created while trying to assure that I did not break the GitTagAction by adding the `RequirePOST` annotation.

File: src/main/java/jenkins/plugins/git/traits/CleanBeforeCheckoutTrait.java [hunk 1]
```diff
@@ -50,6 +50,8 @@ public CleanBeforeCheckoutTrait() {
 
     /**
      * Stapler constructor.
+     *
+     * @param extension the option to clean subdirectories which contain git repositories.
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/traits/CleanBeforeCheckoutTrait.java]: @darxriggs I'm open to a better way to describe this parameter and the same parameter on the other trait.

Comment (darxriggs) [reply src/main/java/jenkins/plugins/git/traits/CleanBeforeCheckoutTrait.java]: If you have a look at the other traits they more or less only link to the type of the parameter. Maybe keep it like this and not duplicate the description of the extension or come up with another description for it.

**Pull request merged**